allennlp/data/iterators/iterator.py

```diff
@@ -0,0 +1,24 @@
+from typing import Dict, List, Generator
+
+import numpy
+
+from ..dataset import Dataset
+from ...common.params import Params
+
+
+class Iterator:
+    """
+    An abstract Iterator class. Iterators must implement __call__, which yields
+    batched examples.
+    """
+    def __call__(self, dataset: Dataset) -> Generator[Dict[str, List[numpy.array]], None, None]:
```

Comment:
I talked with Mark about this, and I think `__call__` is more flexible here - it can take parameters, and can be changed and added to later, while `__iter__` locks you into a particular way of doing things that can't be changed easily.  Also, doing it this way lets us only construct the iterator once, with whatever state is necessary (e.g., batch size, dynamic padding, adaptive batch size...), and use it for all datasets (training, validation, whatever).  The reason we had the ugly `last_num_batches` thing was to ensure that we could keep track of an epoch, and only iterate over a dataset exactly once.  I just added a `yield_one_pass` function to do that instead, as well as a `num_batches_per_epoch` function, if you don't want to use `yield_one_pass`.